src/main/cpp/include/genomicsdb/genomicsdb_columnar_field.h

```diff
@@ -0,0 +1,326 @@
+    /*inline std::vector<size_t>& get_offsets() { return m_offsets; }*/
+    inline size_t* get_offsets_pointer() { return &(m_offsets[0]); }
+    inline size_t get_offsets_size_in_bytes() const { return ((m_offsets.size()-1u)*sizeof(size_t)); } //last offsets element holds the "size"
+    inline size_t get_offsets_length() const { return (m_offsets.size()-1u); } //last offsets element holds the "size"
+    /*
+     * Set offset in buffer of the idx-th element
+     * Useful only for variable length fields and used by the iterators
+     * This allows us to use m_offsets for computing the size of every cell data without
+     * having to write "if(last_cell) else" expressions
+     */
+    inline void set_offset(const size_t idx, const size_t value)
+    {
+      assert(idx < m_offsets.size());
+      m_offsets[idx] = value;
+    }
+    inline size_t get_offset(const size_t idx) const
+    {
+      assert(idx < m_offsets.size() && idx < m_num_filled_entries);
+      return m_offsets[idx ];
+    }
+    inline size_t get_size_of_variable_length_field(const size_t idx) const
+    {
+      assert(idx+1u < m_offsets.size() && idx < m_num_filled_entries);
+      return m_offsets[idx+1u] - m_offsets[idx];
+    }
+    inline bool is_valid(const size_t idx) const
+    {
+      assert(idx < m_valid.size() && idx < m_num_filled_entries);
+      return m_valid[idx];
+    }
+    inline std::vector<bool>& get_valid_vector() { return m_valid; }
+    //Buffer pointers
+    inline void set_next_buffer(GenomicsDBBuffer* next) { m_next_buffer = next; }
+    inline void set_previous_buffer(GenomicsDBBuffer* previous) { m_previous_buffer = previous; }
+    inline GenomicsDBBuffer* get_next_buffer() { return m_next_buffer; }
+    inline GenomicsDBBuffer* get_previous_buffer() { return m_previous_buffer; } 
+  protected:
+    //bool used only in assertions - for debugging
+    bool m_is_in_live_list;
+    std::vector<uint8_t> m_buffer;
+    std::vector<bool> m_valid;
+    //For variable length fields, stores the offsets
+    //If the buffer has data from N-1 cells, this vector has N elements
+    //The total size of the variable length field is always stored in the Nth element
+    //This allows us to use m_offsets for computing the size of every cell data without
+    //having to write "if(last_cell) else" expressions
+    std::vector<size_t> m_offsets;
+    //m_num_filled_entries - the #items filled in the buffer from TileDB
+    //m_num_unprocessed_entries - a counter that's decremented by iterators
+    //m_num_live_entries <= m_num_filled_entries
+    //m_num_unprocessed_entries <= m_num_filled_entries
+    size_t m_num_live_entries;
+    size_t m_num_filled_entries;
+    size_t m_num_unprocessed_entries;
+    //Pointers in the linked list of buffers
+    GenomicsDBBuffer* m_next_buffer;
+    GenomicsDBBuffer* m_previous_buffer;
+};
+
+typedef GenomicsDBBuffer GenomicsDBFixedSizeFieldBuffer;
```

Comment:
Agreed - however, having 2 classes causes a whole lot of trouble. Requires a whole bunch of casts etc.
Since this is a complex, core structure that only experienced devs will be dealing with, I'm ok with leaving this as is for now.